### PR TITLE
Use local_dns_path

### PR DESCRIPTION
**Pull request by Mygod**
Ran out of time today... I will post this WIP here just in case any of you want to continue refactoring the code... :)

### CODE REVIEW

File: src/relay/dnsrelay/upstream.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
+use std::io::Result;
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr, UdpSocket};
+use trust_dns_proto::op::Message;
+
+pub trait Upstream {
+    async fn lookup(&self, request: &Message) -> Result<Message>;
```
Comment (zonyitoo) [inline src/relay/dnsrelay/upstream.rs]: Why your `trait` can have `async fn`? [Shock]

File: src/relay/dnsrelay/upstream.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+impl Upstream for TcpUpstream {
+    async fn lookup(&self, request: &Message) -> Result<Message> {
+        unimplemented!()
+    }
+}
+
+#[cfg(any(target_os = "linux", target_os = "android"))]
+struct UnixSocketUpstream {
+    path: String,
```
Comment (zonyitoo) [inline src/relay/dnsrelay/upstream.rs]: You may want to use `std::path::PathBuf`.

File: src/relay/dnsrelay/upstream.rs (hunk 1 above)
Comment (Mygod) [reply src/relay/dnsrelay/upstream.rs]: I don't know a lot about rust and this is a WIP. At least we can do this in Kotlin. 🤣 

So I guess it should be `fn lookup(&self, request: &Message) -> Future<Result<Message>>`?

Comment (zonyitoo) [reply src/relay/dnsrelay/upstream.rs]: Use a helper macro: `#[async_trait]`

**Comment by madeye** [comment]
`local_dns_path` works now. @Mygod may want to do some refactoring.

**Comment by Mygod** [comment]
Yes... Please do wait for me to catch up...

**Comment by madeye** [comment]
After days of local testing, I think shadowsocks-rust works properly on Android now. No function issue or memory leak observed.

Let's review this change again and hope it get merged soon.

**Comment by zonyitoo** [comment]
LGTM.

**Comment by Mygod** [comment]
I think we can merge this for now and worry about features later. :)

**Pull request merged**